src/couch/src/couch_debug.erl

```diff
@@ -620,6 +676,90 @@ info_size(InfoKV) ->
         {binary, BinInfos} -> lists:sum([S || {_, S, _} <- BinInfos]);
         {_, V} -> V
     end.
+resource_hoggers(MemoryInfo, InfoKey) ->
+    KeyFun = fun
+        ({_Pid, _Id, undefined}) -> undefined;
+        ({_Pid, Id, DataMap}) -> {Id, [{InfoKey, maps:get(InfoKey, DataMap)}]}
+    end,
+    resource_hoggers(MemoryInfo, InfoKey, KeyFun).
+
+resource_hoggers(MemoryInfo, InfoKey, KeyFun) ->
+    HoggersData = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun),
+    TableSpec = [
+        {50, centre, id},
+        {20, centre, InfoKey}
+    ],
+    print_table(HoggersData, TableSpec).
+
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun) when is_atom(InfoKey) ->
+    resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, 20).
+
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, N) when is_atom(InfoKey) and is_integer(N) ->
+    SortedTuples = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, undefined),
+    {TopN, _} = lists:split(N, SortedTuples),
+    TopN;
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, undefined) when is_atom(InfoKey) ->
+    Tuples = lists:filtermap(
+        fun(Tuple) ->
+            case KeyFun(Tuple) of
+                undefined ->
+                    false;
+                Value ->
+                    {true, Value}
+            end
+        end,
+        MemoryInfo
+    ),
+    lists:reverse(lists:keysort(2, Tuples)).
+
+resource_hoggers_snapshot({N, MemoryInfo, InfoKeys} = _Snapshot) ->
+    Data = lists:filtermap(
+        fun({Pid, Id, Data}) ->
+            case memory_info(Pid, InfoKeys) of
+                {Pid, undefined, undefined} ->
+                    false;
+                {_, _, DataMap} ->
+                    {true, {Pid, Id, update_delta(Data, DataMap)}}
+            end
+        end,
+        MemoryInfo
+    ),
+    {N + 1, Data, InfoKeys};
+resource_hoggers_snapshot([]) ->
+    [];
+resource_hoggers_snapshot([{_Pid, _Id, Data} | _] = MemoryInfo) ->
+    resource_hoggers_snapshot({0, MemoryInfo, maps:keys(Data)}).
+
+update_delta({_, InitialDataMap}, DataMap) ->
+    update_delta(InitialDataMap, DataMap);
+update_delta(InitialDataMap, DataMap) ->
+    Delta = maps:fold(
+        fun(Key, Value, AccIn) ->
+            maps:put(Key, maps:get(Key, DataMap, Value) - Value, AccIn)
+        end,
+        maps:new(),
+        InitialDataMap
+    ),
+    {Delta, InitialDataMap}.
+
+analyze_resource_hoggers({N, Data, InfoKeys}, TopN) ->
+    io:format("Number of snapshots: ~p~n", [N]),
+    lists:map(
+        fun(InfoKey) ->
+            KeyFun = fun
+                ({_Pid, _Id, undefined}) -> undefined;
+                ({_Pid, Id, {Delta, _DataMap}}) -> {Id, [{InfoKey, maps:get(InfoKey, Delta)}]}
```

Comment:
In fact maybe there is a way to show both (current and delta). This is what I tried locally:

```erlang
diff --git a/src/couch/src/couch_debug.erl b/src/couch/src/couch_debug.erl
index f93759503..c39cbf5f6 100644
--- a/src/couch/src/couch_debug.erl
+++ b/src/couch/src/couch_debug.erl
@@ -709,13 +709,17 @@ analyze_resource_hoggers({N, Data, InfoKeys}, TopN) ->
         fun(InfoKey) ->
             KeyFun = fun
                 ({_Pid, _Id, undefined}) -> undefined;
-                ({_Pid, Id, {_Delta, DataMap}}) -> {Id, [{InfoKey, maps:get(InfoKey, DataMap)}]}
+                ({_Pid, Id, {Delta, DataMap}}) -> {Id, [
+                    {InfoKey, maps:get(InfoKey, DataMap)},
+                    {delta, maps:get(InfoKey, Delta)}
+                ]}
             end,
             io:format("Top ~p by ~p~n", [TopN, InfoKey]),
-            HoggersData = resource_hoggers_data(Data, InfoKey, KeyFun, TopN),
+            HoggersData = resource_hoggers_data(Data, delta, KeyFun, TopN),
             TableSpec = [
                 {50, centre, id},
-                {20, centre, InfoKey}
+                {20, right, InfoKey},
+                {20, right, delta}
             ],
             print_table(HoggersData, TableSpec)
         end,
```

Unfortunately there is a bug somewhere. Because sorting is done by `InfoKey` not `delta`.

```
MI = couch_debug:memory_info(erlang:processes()).
timer:sleep(3000).
S0 = couch_debug:resource_hoggers_snapshot(MI).
timer:sleep(3000).
S1 = couch_debug:resource_hoggers_snapshot(S0).
couch_debug:analyze_resource_hoggers(S1, 10).
```

```
Top 10 by total_heap_size
|                        id                        |     total_heap_size|               delta
|         couch_prometheus_server[<0.280.0>]       |              150226|              -28690
|                 config[<0.137.0>]                |              139267|                   0
|             erlang:apply/2[<0.1232.0>]           |               86201|              428636
|               code_server[<0.51.0>]              |               86071|                   0
|            couch_log_server[<0.168.0>]           |               36259|                   0
|         couch_stats_aggregator[<0.184.0>]        |               35462|                   0
|             erlang:apply/2[<0.1231.0>]           |               32932|              800170
|  ouch_epi_functions_gen_chttpd_handlers[<0.146.0>|               32875|                   0
|             erl_prim_loader[<0.10.0>]            |               24503|                   0
|     couch_epi_data_gen_flags_config[<0.159.0>]   |               21920|                2585
```

Reply:
The sorting is done by `InfoKey` instead of `Delta` because the sorting occurs in `resource_hoggers_data`: https://github.com/apache/couchdb/pull/3985/files#diff-4806d6fee449d8007647f577455783157f17ce2c886eb66b1fa9f60acb05cc40R713

Reply:
> Unfortunately there is a bug somewhere. Because sorting is done by InfoKey not delta.

This is because `InfoKey` is not used by `resource_hoggers_data`

Reply:
Oh, you assumed there was a `delta` key, there is not.

Reply:
> Oh, you assumed there was a delta key, there is not.

I added it in KeyFun (see diff above)

Reply:
I think we could fix `lists:reverse(lists:keysort(2, Tuples)).` to use passed `InfoKey`.

Reply:
I replaced the `lists:reverse(...` line with the following:

```
    lists:reverse(lists:sort(fun({_, A}, {_, B}) -> %% do we always have tuple with 2 elements?
        lists:keyfind(InfoKey, 1, A) < lists:keyfind(InfoKey, 1, B)
    end, Tuples)).
```

The results are:

```
Top 10 by heap_size
|                        id                        |           heap_size|               delta
|             erlang:apply/2[<0.1657.0>]           |                4185|              314002
|             erlang:apply/2[<0.1658.0>]           |               10958|              185692
|         couch_prometheus_server[<0.280.0>]       |               46422|               28691
|                 chttpd[<0.868.0>]                |                 987|                1599
|            couch_file:init/1[<0.459.0>]          |                 987|                 611
|                   init[<0.0.0>]                  |                1598|                   0
|             erts_code_purger[<0.1.0>]            |                 233|                   0
|    erts_literal_area_collector:start/0[<0.2.0>]  |                 233|                   0
|  rts_dirty_process_signal_handler:start/0[<0.3.0>|                 233|                   0
|  rts_dirty_process_signal_handler:start/0[<0.4.0>|                 233|                   0
```

Reply:
`Top 10 by heap_size` we probably should say something like `Top 10 by change in the heap_size`.

Reply:
You beat me to it. To answer your question, the tuples will always be 2 elements - an InfoKey and a value.

Reply:
Tested it locally:
```

(node1@127.0.0.1)6> couch_debug:analyze_resource_hoggers(RHS, 3).
Number of snapshots: 1
Top 3 by change in binary
|                        id                        |              binary|               delta
|             erl_prim_loader[<0.10.0>]            |               34373|               27493
|                   init[<0.0.0>]                  |                   0|                   0
|             erts_code_purger[<0.1.0>]            |                   0|                   0
Top 3 by change in dictionary
|                        id                        |          dictionary|               delta
|             erlang:apply/2[<0.1205.0>]           |                   1|                   4
|                   init[<0.0.0>]                  |                   0|                   0
|             erts_code_purger[<0.1.0>]            |                   0|                   0
Top 3 by change in heap_size
|                        id                        |           heap_size|               delta
|             erlang:apply/2[<0.1205.0>]           |                4185|              192465
|      couch_replicator_rate_limiter[<0.517.0>]    |                 233|                1365
|                   init[<0.0.0>]                  |                2586|                   0
Top 3 by change in links
|                        id                        |               links|               delta
|                   init[<0.0.0>]                  |                   3|                   0
|             erts_code_purger[<0.1.0>]            |                   0|                   0
|    erts_literal_area_collector:start/0[<0.2.0>]  |                   0|                   0
Top 3 by change in memory
|                        id                        |              memory|               delta
|             erlang:apply/2[<0.1205.0>]           |              264040|             1311072
|      couch_replicator_rate_limiter[<0.517.0>]    |                2832|               31608
|    folsom_sample_slide_server:init/1[<0.235.0>]  |                2752|                3008
Top 3 by change in message_queue_len
|                        id                        |   message_queue_len|               delta
|                   init[<0.0.0>]                  |                   0|                   0
|             erts_code_purger[<0.1.0>]            |                   0|                   0
|    erts_literal_area_collector:start/0[<0.2.0>]  |                   0|                   0
Top 3 by change in monitored_by
|                        id                        |        monitored_by|               delta
|                   init[<0.0.0>]                  |                   0|                   0
|             erts_code_purger[<0.1.0>]            |                   0|                   0
|    erts_literal_area_collector:start/0[<0.2.0>]  |                   0|                   0
Top 3 by change in monitors
|                        id                        |            monitors|               delta
|                   init[<0.0.0>]                  |                   0|                   0
|             erts_code_purger[<0.1.0>]            |                   0|                   0
|    erts_literal_area_collector:start/0[<0.2.0>]  |                   0|                   0
Top 3 by change in stack_size
|                        id                        |          stack_size|               delta
|                   init[<0.0.0>]                  |                   3|                   0
|             erts_code_purger[<0.1.0>]            |                   2|                   0
|    erts_literal_area_collector:start/0[<0.2.0>]  |                   6|                   0
Top 3 by change in total_heap_size
|                        id                        |     total_heap_size|               delta
|             erlang:apply/2[<0.1205.0>]           |               32875|              163884
|      couch_replicator_rate_limiter[<0.517.0>]    |                 233|                3951
|    folsom_sample_slide_server:init/1[<0.235.0>]  |                 233|                 376
```